flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayIntersectFunction.java

```diff
@@ -0,0 +1,156 @@
+import java.util.List;
+import java.util.Set;
+
+import static org.apache.flink.table.api.Expressions.$;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#ARRAY_INTERSECT}. */
+@Internal
+public class ArrayIntersectFunction extends BuiltInScalarFunction {
+    private final ArrayData.ElementGetter elementGetter;
+    private final SpecializedFunction.ExpressionEvaluator hashcodeEvaluator;
+    private final SpecializedFunction.ExpressionEvaluator equalityEvaluator;
+    private transient MethodHandle hashcodeHandle;
+    private transient MethodHandle equalityHandle;
+
+    public ArrayIntersectFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.ARRAY_INTERSECT, context);
+        final DataType dataType =
+                ((CollectionDataType) context.getCallContext().getArgumentDataTypes().get(0))
+                        .getElementDataType()
+                        .toInternal();
+        elementGetter = ArrayData.createElementGetter(dataType.toInternal().getLogicalType());
+        hashcodeEvaluator =
+                context.createEvaluator(
+                        Expressions.call("$HASHCODE$1", $("element1")),
+                        DataTypes.INT(),
+                        DataTypes.FIELD("element1", dataType.notNull().toInternal()));
+        equalityEvaluator =
+                context.createEvaluator(
+                        $("element1").isEqual($("element2")),
+                        DataTypes.BOOLEAN(),
+                        DataTypes.FIELD("element1", dataType.notNull().toInternal()),
+                        DataTypes.FIELD("element2", dataType.notNull().toInternal()));
+    }
+
+    @Override
+    public void open(FunctionContext context) throws Exception {
+        hashcodeHandle = hashcodeEvaluator.open(context);
+        equalityHandle = equalityEvaluator.open(context);
+    }
+
+    public @Nullable ArrayData eval(ArrayData array1, ArrayData array2) {
+        try {
+            if (array1 == null || array2 == null) {
+                return null;
+            }
+            List<Object> list = new ArrayList<>();
+            Set<ObjectContainer> seen = new HashSet<>();
+
+            boolean isNullPresentInArrayTwo = false;
+            for (int pos = 0; pos < array2.size(); pos++) {
+                final Object element = elementGetter.getElementOrNull(array2, pos);
+                if (element == null) {
+                    isNullPresentInArrayTwo = true;
+                } else {
+                    ObjectContainer objectContainer = new ObjectContainer(element);
+                    seen.add(objectContainer);
+                }
+            }
+            boolean isNullPresentInArrayOne = false;
+            for (int pos = 0; pos < array1.size(); pos++) {
+                final Object element = elementGetter.getElementOrNull(array1, pos);
+                if (element == null) {
+                    isNullPresentInArrayOne = true;
```

Comment:
Can't we just check if it was present in the second array and add it to the `list`?

Reply:
Do you mean NULL? We need to check whether both array1 and array2 contain NULL. If both contain NULL, we can include it in the list.

The description of array_intersect is like this
`Returns an array of the elements in the intersection of array1 and array2, without duplicates.`